presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1170,6 +1152,72 @@ else if (criteria instanceof JoinOn) {
             return output;
         }
 
+        private Scope analyzeJoinUsing(Join node, List<Identifier> columns, Optional<Scope> scope, Scope left, Scope right)
+        {
+            List<Field> joinFields = new ArrayList<>();
+
+            List<Integer> leftJoinFields = new ArrayList<>();
+            List<Integer> rightJoinFields = new ArrayList<>();
+
+            Set<Identifier> seen = new HashSet<>();
+            for (Identifier column : columns) {
+                if (!seen.add(column)) {
+                    throw new SemanticException(DUPLICATE_COLUMN_NAME, column, "Column '%s' appears multiple times in USING clause", column.getValue());
+                }
+
+                QualifiedName columnName = QualifiedName.of(column.getValue());
+                Optional<ResolvedField> leftField = left.tryResolveField(column, columnName);
+                Optional<ResolvedField> rightField = right.tryResolveField(column, columnName);
+
+                if (!leftField.isPresent()) {
+                    throw new SemanticException(MISSING_ATTRIBUTE, column, "Column '%s' is missing from left side of join", columnName);
+                }
+                if (!rightField.isPresent()) {
+                    throw new SemanticException(MISSING_ATTRIBUTE, column, "Column '%s' is missing from right side of join", columnName);
+                }
+
+                // ensure a comparison operator exists for the given types (applying coercions if necessary)
+                try {
+                    metadata.getFunctionRegistry().resolveOperator(OperatorType.EQUAL, ImmutableList.of(
```

Comment:
It's hard to report accurate errors that way. If a field is missing or there's a type mismatch, it would be reported for the synthetic expression, no the terms in the original SQL tree.

Also, FieldReference is actually something that shouldn't exist in the SQL tree. It's just an artifact of reusing the SQL tree for the IR (which I want to move away from).

Reply:
> It's hard to report accurate errors that way. If a field is missing or there's a type mismatch, it would be reported for the synthetic expression, no the terms in the original SQL tree.

True. `FieldReference` maybe could have `NodeLocation`, but that seems like a hack. 

> It's just an artifact of reusing the SQL tree for the IR (which I want to move away from).

I think `FieldReference` is used in a dual way. Once for channel referencing (during physical planning) and once during `analysis<->planning` to reference anonymous columns (e.g '*' columns) or referencing column by long literal.
In the latter case maybe we could replace `FieldReference` by some kind of temporary identifier? For instance:
```
SELECT a ORDER BY 1
```
`1` gets converted to `FieldReference`, but could be converted to some temporary identifier. I don't think `FieldReference` is that evil in this context anyway.